internal/graph/limited_checker.go

```diff
@@ -0,0 +1,60 @@
+package graph
+
+import (
+	"context"
+	"fmt"
+)
+
+type LimitedLocalChecker struct {
+	resolver CheckResolver
+	limiter  chan struct{}
+}
+
+type LimitedLocalCheckerOpt func(l *LimitedLocalChecker)
+
+func WithCheckResolver(resolver CheckResolver) LimitedLocalCheckerOpt {
+	return func(l *LimitedLocalChecker) {
+		l.resolver = resolver
+	}
+}
+
+func WithConcurrencyLimit(limit uint32) LimitedLocalCheckerOpt {
+	return func(l *LimitedLocalChecker) {
+		l.limiter = make(chan struct{}, limit)
+	}
+}
+
+func MustNewLimitedLocalChecker(opts ...LimitedLocalCheckerOpt) *LimitedLocalChecker {
+	checker, err := NewLimitedLocalChecker(opts...)
+	if err != nil {
+		panic(err)
+	}
+
+	return checker
+}
+
+func NewLimitedLocalChecker(opts ...LimitedLocalCheckerOpt) (*LimitedLocalChecker, error) {
+
+	checker := &LimitedLocalChecker{
+		limiter: make(chan struct{}, 50),
+	}
+
+	for _, opt := range opts {
+		opt(checker)
+	}
+
+	if checker.resolver == nil {
+		return nil, fmt.Errorf("the provided CheckResolver cannot be nil")
+	}
+
+	return checker, nil
+}
+
+func (l *LimitedLocalChecker) ResolveCheck(ctx context.Context, req *ResolveCheckRequest) (*ResolveCheckResponse, error) {
+	l.limiter <- struct{}{}
+	defer func() {
+		<-l.limiter
+	}()
```

Comment:
This will rate-limit `LimitedLocalChecker.ResolveCheck`, but not `LocalChecker.ResolveCheck`, which is the one that is called recursively

I did a quick load test lasting 1 minute with 100 concurrent Checks, this is the number of goroutines:

<img width="1916" alt="image" src="https://github.com/openfga/openfga/assets/5374887/ccf07ef1-4a5d-49b7-9822-457e9d778269">

Reply:
@miparnisari try it now. I hadn't correctly wired up the CheckResolver so that it would loop back. It should now be limited properly.